proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,1028 @@
+
+`xdstp://[{authority}]/{resource type}/{id/*}?{context parameters}{#processing directive,*}`
+
+* `authority` is an opaque string naming a resource authority, e.g.
+  `some.control.plane`.
+* `resource type` is the xDS resource type URL, e.g. `envoy.config.route.v3.RouteConfiguration`
+* `id/*` is the remaining path component of the URI and is fully opaque; naming
+  is at the discretion of the control plane(s).
+* `context parameters` are the URI query parameters and express contextual
+  information for selecting resource variants, for example `shard_id=1234&direction=inbound`.
+* A `processing directive` provides additional information to the client on how
+  the URI is to be interpreted. Two supported directives are:
+  * `alt=<xdstp:// URI>` expressing a fallback URI.
+  * `entry=<resource name>` providing an anchor referring to an inlined resource
+    for [list collections](#list). Resource names must be of the form `[a-zA-Z0-9_-\./]+`.
+
+Multiple directives are comma separated in the fragment component of the URI.
+
+An example application of the above URI scheme is:
+
+`xdstp://some.control.plane/envoy.config.route.v3.RouteConfiguration/foo/bar?shard_id=1234&direction=inbound`
+
+We distinguish between two forms of `xdstp://` URIs:
+
+* Uniform Resource Locators (URLs) instructing a client on how to locate a given
+  xDS resource. URLs may include processing directives.
+
+* Uniform Resource Names (URNs) referring to a specific xDS resource. URNs do not
+  include processing directives. `xdstp://` URNs provide self-contained cache keys.
+
+#### Proto3 expression
+
+`xdstp://` URIs are canonically encoded in `proto3` messages. The structure
+for this encoding is provided at https://github.com/cncf/xds/tree/main/xds/core/v3:
+
+* [URL](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_locator.proto) definition
+* [URN](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_name.proto) definition
+
+#### Normalization
+
+Two `xdstp://` URNs are considered equivalent if they match component-wise
+modulo context parameter ordering.
+
+## Context parameters
+
+Context parameters in URNs presented by the client to the server will be dervied
+by the following process. Using an example of a URL
+`xdstp://some-authority/some.type/foo?xds.bar=baz`:
+
+1. Static `Node`-derived context parameters are first added. These are prefixed with
+   `xds.node.`. The set of `Node`-derived context parameters is specified in the
+   bootstrap on a per-resource type basis. The key reflects the `Node` proto3
+   field structure, e.g. `xds.node.locality.sub_zone=some_sub_zone`,
+   `xds.node.metadata.bar="a"`, `xds.node.user_agent_version=1.2.3`.
+
+2. Dynamic context parameters are added next. These are free form and at the
+   discretion of the client. They can provide client-specific fields such as
+   `xds.shard_id=1234` and potentially change through the lifetime of an xDS
+   stream.
```

Comment:
The context parameter changes in the client. This then causes the client to ask for a new request from the management server. The management server only observes a new request with an `xdstp://` URL with different context parameters, caching still continues to work at the management server, since the request now refers to a different resource.

Reply:
Yes that makes sense. Rephrase? It sounds like this can change mid-stream.

Reply:
Updated, LMK WDYT.